### PR TITLE
[afreecatv] Support catchstory

### LINKED ISSUE #10112: Afreecatv.com catchstory
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Albania

### Example URLs

https://vod.afreecatv.com/player/103247/catchstory 

### Provide a description that is worded well enough to be understood

catchstory is a replay of the older videos with story behind it, usually contains multiple videos. Its possible to download them via grabbing m3u8 index. Thanks

```python
yt-dlp --no-part --restrict-filenames --output "[캐치]완전 큰 꼬르륵 _ AfreecaTV VOD.%(ext)s" "https://vod-normal-global-cdn-z02.afreecatv.com/v101/hls/clip//605/4c857f9e-078b-4e7c-aad6-8dd6653df2bf/4201605.smil/seek_start_ms=0/duration_ms=16000/original/both/playlist.m3u8"
``` 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
yt-dlp --legacy-server-connect https://vod.afreecatv.com/player/103247/catchstory -v
[debug] Command-line config: ['--legacy-server-connect', 'https://vod.afreecatv.com/player/103247/catchstory', '-v']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.05.27 from yt-dlp/yt-dlp [12b248ce6] (pip)
[debug] Python 3.10.10 (CPython AMD64 64bit) - Windows-10-10.0.17763-SP0 (OpenSSL 1.1.1s  1 Nov 2022)
[debug] exe versions: ffmpeg N-114162-g2129d66a66-20240314 (setts), ffprobe N-114162-g2129d66a66-20240314
[debug] Optional libraries: Cryptodome-3.20.0, brotli-1.1.0, certifi-2024.02.02, mutagen-1.47.0, requests-2.31.0, sqlite3-3.39.4, urllib3-2.2.1, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1820 extractors
[afreecatv] Extracting URL: https://vod.afreecatv.com/player/103247/catchstory
[afreecatv] 103247: Downloading JSON metadata
ERROR: [afreecatv] 103247: The VOD does not exist
```

### PR DESCRIPTION
**Pull request by hui1601**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Support afreecaTv catch-story(similar to YouTube Shorts)

Closes #10112

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [X] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by PROFESSOR-SOURABH** [comment]
Thanks

File: yt_dlp/extractor/afreecatv.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ class AfreecaTVIE(AfreecaTVBaseIE):
                             )\?.*?\bnTitleNo=|
                             vod\.afreecatv\.com/(PLAYER/STATION|player)/
                         )
-                        (?P<id>\d+)
+                        (?P<id>\d+)[^/]*$
```
Comment (bashonly) [inline yt_dlp/extractor/afreecatv.py:75]:
```suggestion
                        (?P<id>\d+)/?(?:$|[?#&])
```

File: yt_dlp/extractor/afreecatv.py [hunk 2]
```diff
@@ -253,6 +253,50 @@ def _real_extract(self, url):
...
+            'id': '103247',
+        },
+        'playlist_count': 2,
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        data = self._download_json(
+            'http://api.m.afreecatv.com/catchstory/a/view', video_id, headers={'Referer': url},
+            query={'aStoryListIdx': '', 'nStoryIdx': video_id}).get('data', [])
```
Comment (bashonly) [inline yt_dlp/extractor/afreecatv.py:273]: see https://github.com/yt-dlp/yt-dlp/pull/10174#issuecomment-2195320898 about using `impersonate` instead of insecure requests

also, let's use `traverse_obj` below when we iterate over the data instead of `.get()` here
```suggestion
        data = self._download_json(
            'https://api.m.afreecatv.com/catchstory/a/view', video_id, headers={'Referer': url},
            query={'aStoryListIdx': '', 'nStoryIdx': video_id}, impersonate=True)
```

File: yt_dlp/extractor/afreecatv.py [hunk 3]
```diff
@@ -253,6 +253,50 @@ def _real_extract(self, url):
...
 
 
+class AfreecaTVCatchStoryIE(AfreecaTVBaseIE):
+    IE_NAME = 'afreecatv:catchstory'
+    IE_DESC = 'afreecatv.com catch story'
+    _VALID_URL = r'https?://vod\.afreecatv\.com/player/(?P<id>\d+)/catchstory'
+    _TESTS = [{
+        'url': 'https://vod.afreecatv.com/player/103247/catchstory',
+        'info_dict': {
+            '_type': 'playlist',
```
Comment (bashonly) [inline yt_dlp/extractor/afreecatv.py:263]:
```suggestion
```

File: yt_dlp/extractor/afreecatv.py [hunk 4]
```diff
@@ -253,6 +253,50 @@ def _real_extract(self, url):
...
+                    'uploader_id': catch.get('writer_id'),
+                    'thumbnail': catch.get('thumb'),
+                    'duration': file.get('duration'),
+                    'formats': [{
+                        'url': file.get('file'),
+                        'format_id': 'http',
+                    }],
+                    'timestamp': file.get('write_timestamp'),
+                })
+        return self.playlist_result(entries, video_id)
```
Comment (bashonly) [inline yt_dlp/extractor/afreecatv.py:297]: - avoid append loops for entries, use a generator/PagedList when possible
- we can use traversal to filter for what we need, check value types, and transform values
- `write_timestamp` is found in `catch`, not `file`
- `duration` is in milliseconds, we need to convert
- the URLs are m3u8s, so `http` is not correct as `format_id`
- we can just return `url` in the info dict root instead of a `formats` field
- we can move the entries generator to a static method

```suggestion

        return self.playlist_result(self._entries(data), video_id)

    @staticmethod
    def _entries(data):
        # 'files' is always a list with 1 element
        yield from traverse_obj(data, (
            'data', lambda _, v: v['story_type'] == 'catch',
            'catch_list', lambda _, v: v['files'][0]['file'], {
                'id': ('files', 0, 'file_info_key', {str}),
                'url': ('files', 0, 'file', {url_or_none}),
                'duration': ('files', 0, 'duration', {functools.partial(int_or_none, scale=1000)}),
                'title': ('title', {str}),
                'uploader': ('writer_nick', {str}),
                'uploader_id': ('writer_id', {str}),
                'thumbnail': ('thumb', {url_or_none}),
                'timestamp': ('write_timestamp', {int_or_none}),
            }))
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/afreecatv.py [hunk 5]
```diff
@@ -253,6 +253,44 @@ def _real_extract(self, url):
...
+
+    @staticmethod
+    def _entries(data):
+        # 'files' is always a list with 1 element
+        yield from traverse_obj(data, (
+            'data', lambda _, v: v['story_type'] == 'catch',
+            'catch_list', lambda _, v: v['files'][0]['file'], {
+                'id': ('files', 0, 'file_info_key', {str}),
+                'url': ('files', 0, 'file', {url_or_none}),
+                'ext': ('src_movie', {lambda v: determine_ext(v, 'mp4')}),
```
Comment (bashonly) [inline yt_dlp/extractor/afreecatv.py:284]: we don't want this. the test's url is an m3u8 and this value is mp4. it will not download correctly if yt-dlp thinks an m3u8 is an mp4
```suggestion
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/afreecatv.py (hunk 5 above)
Comment (hui1601) [reply yt_dlp/extractor/afreecatv.py:284]: I don't understand why, but sometimes it has the m3u8 extension but is actually mp4. Since this problem only occurs in a very few cases, I will revert that commit.

Comment (bashonly) [reply yt_dlp/extractor/afreecatv.py:284]: No, it's an m3u8, and yt-dlp downloads it with its HLS downloader and muxes it to mp4

Comment (bashonly) [reply yt_dlp/extractor/afreecatv.py]: if there are other catchstory links where the format is actually an mp4 could we add a test for one of those? btw do these stories expire?

**Comment by hui1601** [comment]
I'm sorry I bothered you.
I have a question. AfreecaTv is scheduled to change to sooplive.co.kr around October.
Then, can I create a PR to change the existing extractor name? Or am I wondering if I should keep the existing extractor and create a new extractor.

**Comment by bashonly** [comment]
> I'm sorry I bothered you.

No bother at all. It's all good.

> I have a question. AfreecaTv is scheduled to change to sooplive.co.kr around October.
> Then, can I create a PR to change the existing extractor name? Or am I wondering if I should keep the existing extractor and create a new extractor.

We typically try to avoid changing the extractor class names. We *could* change the `IE_NAME` value(s), but that would only be appropriate if the video `id`s are going to carry over from afreecatv to sooplive. If not, then we should just write new extractor(s)

File: yt_dlp/extractor/afreecatv.py (hunk 5 above)
Comment (hui1601) [reply yt_dlp/extractor/afreecatv.py]: Because it is a similar service to YouTube's Shorts, it does not seem to be deleted. But I don't think there's a need to deal with edge cases.
In this case, it was uploaded from a dummy account. I think it was uploaded for development testing.

Comment (bashonly) [reply yt_dlp/extractor/afreecatv.py]: > But I don't think there's a need to deal with edge cases.

If you have a link where the format url is a direct mp4 then we should add it; test coverage is good

**Review by bashonly — APPROVED** [review]
